Upload video

DinhAnh-Nguyen opened the pull request:
## Why was this PR opened?

## Brief Description

## Unrelated Changes

## UI Screenshots

## Link to Referenced Trello Cards

## Mentions

DinhAnh-Nguyen committed:
feat: adding video attribute for techDataSchema

DinhAnh-Nguyen committed:
feat: video upload logic function added

DinhAnh-Nguyen committed:
feat: displaying the uploaded video on myProfile page

kwoodman1970 reviewed (changes requested):
There are some minor issues with naming conventions and styles, but there are bigger issues to consider.

There needs to be a progress indicator for uploading a video.

What happens if a user navigates away from the "Update Profile" page before the video has finished being uploaded?

When a user changes their video, what happens to the old video file?

How can a user delete their video?

If I add a video, then navigate to the "My Profile" page, then navigate back to the "Update Profile" page, the video isn't there.

We might have to give some thought regarding the placement of this video on these two components.

DinhAnh-Nguyen committed:
fix: changing variable name newVideo to videoFile

DinhAnh-Nguyen committed:
fix: changing function name from handleButtonClick to openVideoFilePicker

DinhAnh-Nguyen committed:
fix: remove changeEvent import

DinhAnh-Nguyen committed:
fix: moving ContractorProfile stylings to the css file

DinhAnh-Nguyen committed:
fix: moving all ProfileForm inline layout to ProfileForm.module.css

DinhAnh-Nguyen committed:
fix: merging two routines into one function uploadFileAndGetUrl

DinhAnh-Nguyen committed:
fix: allowing current profile image to show as default user avatar in Update Profile page

DinhAnh-Nguyen committed:
fix: changing video tag property name from autoplay to autoPlay

DinhAnh-Nguyen committed:
fix: allowing updated video to be shown on user Update Profile page

DinhAnh-Nguyen committed:
feat: adding autoPlay feature to the video tag on My Profile page

DinhAnh-Nguyen committed:
feat: add function to remove old video file when new video is uploaded

Creating one single removeOldFile function that works for both images and videos

DinhAnh-Nguyen committed:
feat: adding delete video function

DinhAnh-Nguyen committed:
fix: remove unused hook

DinhAnh-Nguyen committed:
feat: adding a progress indicator for uploading videos

DinhAnh-Nguyen committed:
feat: blocking users from making further actions when the videos are uploading

kwoodman1970 committed:
Merge branch 'main' into Daniel--Upload-Video

kwoodman1970 commented on src/components/ContractorProfile/ContractorProfile.jsx:
```diff
@@ -132,7 +132,7 @@ const ContractorProfile = (props) => {
           <section>
             <h2>Video</h2>
-            <video src={contractor.video} controls style={{ width: '300px', marginTop: '10px' }} />
+            <video src={contractor.video} controls autoPlay />
```
It may not be a good idea to make videos to play automatically.  First, it can consume a lot of bandwidth.  Second, it can have a jarring effect on someone who's trying to quickly read contractors' "About" sections before delving deeper.

*We can discuss this at the next online meeting.*

DinhAnh-Nguyen replied:
@kwoodman1970 I apologize since I won't be able to attend any future meetings due to my new schedule. 

I think you've made a really good point. Should we just remove the autoPlay feature?

kwoodman1970 replied:
Yes, I think we should remove it.

kwoodman1970 commented on src/components/ContractorProfile/ContractorProfile.module.css:
```diff
@@ -47,4 +47,9 @@
 	cursor:  pointer
+}
+
+video {
```
HTML selectors can affect every page in a website, even if they're inside a CSS module.  CSS modules should only have CSS class selectors (although HTML elements can be child selectors of CSS classes) to ensure that their effects are restricted to the page that imported them.

kwoodman1970 commented on src/components/profileForm/ProfileForm.jsx:
```diff
@@ -13,19 +13,60 @@ import DeleteAccount from '../DeleteAccount';
 import { useNavigate } from 'react-router-dom';
 import ResponsiveGrid from '../ResponsiveGrid';
 import Badge from '../Badge';
+import { UNSAFE_NavigationContext as NavigationContext } from 'react-router-dom';
```
Anything that's explicitly marked "unsafe" is a red flag.  Usually there's a good reason for such designations – for example, there may be memory leaks or security vulnerabilities.

Is there a "safe" alternative that will serve our purposes?

kwoodman1970 commented on src/components/profileForm/ProfileForm.jsx:
```diff
@@ -13,19 +13,60 @@ import DeleteAccount from '../DeleteAccount';
 	const [projects, setProjects] = useState(initialFormData.projects);
 
+	useEffect(() => {
+        const handleBeforeUnload = (e) => {
```
Eschew single-letter names.  `event` would be a better name for this parameter.

kwoodman1970 commented on src/components/profileForm/ProfileForm.jsx:
```diff
@@ -13,19 +13,60 @@ import DeleteAccount from '../DeleteAccount';
+        return () => {
+            window.removeEventListener('beforeunload', handleBeforeUnload);
+        };
+    }, [uploadProgress]);
```
The indentation in this `useEffect()` block appears to be four columns using spaces, which is inconsistent with the rest of the file (also, it isn't done correctly).  Please change this to eight columns using tabs.

kwoodman1970 commented on src/components/profileForm/ProfileForm.jsx:
```diff
@@ -13,19 +13,60 @@ import DeleteAccount from '../DeleteAccount';
+            navigator.replace = originalReplace;
+        };
+    }, [shouldBlock, message, navigator]);
+}
```
The indentation in this function appears to be four columns using spaces, which is inconsistent with the rest of the file.  Please change this to eight columns using tabs.

kwoodman1970 commented on src/components/profileForm/ProfileForm.jsx:
```diff
@@ -58,46 +99,105 @@ export default function ProfileForm(props) {
+				toast.error('Failed to upload video.');
+				setUploadProgress(0);
+			}
 		}
```
This is probably not the best point at which to actually upload video files.  A user could upload several in a row, then navigate away from this page without saving.  That would result in orphaned files in the file store.  It would be better to upload a new video file when the user clicks the `Save` button.

kwoodman1970 commented on src/components/profileForm/ProfileForm.jsx:
```diff
@@ -13,19 +13,60 @@ import DeleteAccount from '../DeleteAccount';
+                e.preventDefault();
+                e.returnValue = '';
+            }
+        };
```
What happens if `uploadProgress` is 0?  Shouldn't that be handled the same way as if it were otherwise less than 100?

kwoodman1970 commented on src/components/profileForm/ProfileForm.module.css:
```diff
@@ -35,3 +40,88 @@ contained in it will be spread out evenly across it.
+}
+
+.ProgressBar {
+  background: #4caf50;
```
We should try to use one of the colours in the colour palette from the visual design (see the `src/index.css` file) in order to maintain a consistent appearance.

kwoodman1970 commented on src/components/profileForm/ProfileForm.module.css:
```diff
@@ -35,3 +40,88 @@ contained in it will be spread out evenly across it.
+.ProgressBarWrapper {
+  margin: 10px 0;
+  width: 100%;
+  background: #eee;
```
We should try to use one of the colours in the colour palette from the visual design (see the `src/index.css` file) in order to maintain a consistent appearance.

kwoodman1970 commented on src/components/profileForm/ProfileForm.jsx:
```diff
@@ -170,19 +270,33 @@ export default function ProfileForm(props) {
+									<div className={styles.ProgressBarWrapper}>
+										<div className={styles.ProgressBar} style={{ width: `${uploadProgress}%` }} />
+										<span>{uploadProgress}%</span>
+									</div>
```
This progress bar is a prime candidate for a refactoring into a separate React component.

kwoodman1970 reviewed (changes requested):
This is turning out very nicely!

DinhAnh-Nguyen committed:
fix: adding classname for video component

DinhAnh-Nguyen committed:
fix: Remove autoPlay property from video component

DinhAnh-Nguyen committed:
fix: Improve code

- Changed function parameter from e to event
- Removed unused modules and prop values
